### PR TITLE
HPCC-20100 Limit reply size by configurable option in request

### PR DESCRIPTION
**Pull request by jakesmith**
"replyLimit" in request defines the limit in KB (default 1024).

Also:
1) move format, new replyLimit and activity into a
CRemoteRequest object.
2) Move common request attributes into base (just chooseN for now)

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@rpastrana - please review and test for performance impact.

File: common/remote/sockfile.cpp [hunk 1]
```diff
@@ -3984,8 +3999,9 @@ class CRemoteDiskBaseActivity : public CSimpleInterfaceOf<IRemoteActivity>, impl
 public:
     IMPLEMENT_IINTERFACE_USING(CSimpleInterfaceOf<IRemoteActivity>)
 
-    CRemoteDiskBaseActivity()
+    CRemoteDiskBaseActivity(IPropertyTree &config)
```
Comment (rpastrana) [inline common/remote/sockfile.cpp]: why not IPropertyTree * config

File: common/remote/sockfile.cpp [hunk 2]
```diff
@@ -4110,12 +4125,11 @@ class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
             return true;
     }
 public:
-    CRemoteDiskReadActivity(IPropertyTree &config) : prefetchBuffer(nullptr)
+    CRemoteDiskReadActivity(IPropertyTree &config) : PARENT(config), prefetchBuffer(nullptr)
```
Comment (rpastrana) [inline common/remote/sockfile.cpp]: PARENT here is just a typedef of the base class? if so is this purely stylistic or are we gaining something

File: common/remote/sockfile.cpp [hunk 3]
```diff
@@ -6227,6 +6241,7 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
 
         int cursorHandle = requestTree->getPropInt("handle");
         OutputFormat outputFormat = outFmt_Xml;
+        unsigned __int64 replyLimit = 0;
```
Comment (rpastrana) [inline common/remote/sockfile.cpp]: does it make sense to initialize to the default value here as well?

File: common/remote/sockfile.cpp [hunk 4]
```diff
@@ -6330,10 +6351,11 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
                         }
                         pastFirstRow = true;
                     }
+                    while (reply.length() < replyLimit);
```
Comment (rpastrana) [inline common/remote/sockfile.cpp]: do we ever stream out partial records?

**Review by rpastrana — COMMENTED** [review]
@jakesmith just a couple of comments/question

File: common/remote/sockfile.cpp (hunk 1 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: + consistency with existing derivatives
+ when it is illegal for object/interface to be null, reference is clearer.
+ preference when purely used and not owned (not link counted etc).

File: common/remote/sockfile.cpp (hunk 2 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: yes, just a typedef.
It makes life easier when refactoring that PARENT is always just that - the parent of the current class, e.g. for all the overloaded methods that call back up to parent class.
There aren't many use cases here, but generally in code else where you'll see code like:
```
    virtual void someFunc() override
    {
         PARENT::someFunc();
         dootherstuff..
    }
```

File: common/remote/sockfile.cpp (hunk 3 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: yes I think so, it's clearer, the compiler will likely optimize it away anyway. And makes sure all code paths are initialized even if current ones where not used don't matter.

File: common/remote/sockfile.cpp (hunk 4 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: no.

File: common/remote/sockfile.cpp (hunk 3 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Ah you meant to : requestTree->getPropInt64("replyLimit", defaultDaFSReplyLimitKB) * 1024

Could do, and if zero cost perhaps, but as no practical difference and there's extra expense of going through another call to getPropInt virtual etc, on balance better as it is.

Comment (rpastrana) [reply common/remote/sockfile.cpp]: Yes, I think one of my comments didn't go through, I think the chooseN is initialized the same way. 
Anyway,the only extra expense I see in this case would be the second "* 1024". Not a big deal

**Review by rpastrana — APPROVED** [review]
@jakesmith approved

**Comment by jakesmith** [comment]
One of the smoke tests failed - investigating.

**Comment by rpastrana** [comment]
@jakesmith just to clarify, the initial;cursor;handle requests will need to specify the reply limit, correct?

**Comment by jakesmith** [comment]
Yes, only the initial message, not the continuation messages.

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

File: common/remote/sockfile.cpp (hunk 4 above)
Comment (richardkchapman) [reply common/remote/sockfile.cpp]: Do we cope with case where single record is larger than the limit?

Comment (jakesmith) [reply common/remote/sockfile.cpp]: Yes in the sense that at least 1 record is serialized regardless, so if 1 rec is > than limit, then the returned size will be approx whatever that 1 rec size is.

**Comment by jakesmith** [comment]
@richardkchapman - replied above.

File: common/remote/sockfile.cpp (hunk 4 above)
Comment (richardkchapman) [reply common/remote/sockfile.cpp]: Ok - caller / API documentation will need to be aware that the passed-in size limit may be exceeded in this case then.

**Pull request merged**